compiler/luci/pass/src/FoldDequantizePass.cpp

```diff
@@ -0,0 +1,190 @@
+      if (qd >= const_node->quantparam()->zerop.size())
+        qd = 0;
+
+      new_const_node->at<loco::DataType::FLOAT32>(i) =
+          (float)(const_node->at<loco::DataType::S8>(i) - const_node->quantparam()->zerop.at(qd)) *
+          const_node->quantparam()->scale.at(qd);
+    }
+  }
+  else
+  {
+    for (uint32_t i = 0; i < const_node->size<loco::DataType::U8>(); ++i)
+    {
+      uint32_t qd = (i % right_count) / (right_count / q_dim_value);
+      if (qd >= const_node->quantparam()->zerop.size())
+        qd = 0;
+
+      new_const_node->at<loco::DataType::FLOAT32>(i) =
+          (float)((int)const_node->at<loco::DataType::U8>(i) -
+                  const_node->quantparam()->zerop.at(qd)) *
+          const_node->quantparam()->scale.at(qd);
+    }
+  }
+
+  return new_const_node;
+}
+
+} // namespace
+
+namespace luci
+{
+
+/**
+ * Folding pattern 1 - When input of Dequantize is constant
+ *
+ * [Before]
+ *     quantized_const_input ----- Dequantize ----- Op1 ---
+ *                             +-- Op2 ---
+ *                             +-- Op3 ---
+ *
+ * [After]
+ *   dequantized_const_input ---------------------- Op1 ---
+ *
+ *     quantized_const_input ----- Op2 ---
+ *                             +-- Op3 ---
+ *
+ *
+ * Folding pattern 2 - When input of Dequantize is using quantized value
+ *
+ * [Before]
+ *     quantized_const_input ----- Gather ------ Dequantize --- Op1 ---
+ *                             +-- Op2 ---
+ *                             +-- Op3 ---
+ *
+ * [After]
+ *   dequantized_const_input ------Gather ------ Op1 ---
+ *
+ *     quantized_const_input ----- Op2 ---
+ *                             +-- Op3 ---
+ *
+ *
+ */
+bool FoldDequantizePass::run(loco::Graph *g)
+{
+  bool changed = false;
+
+  for (auto node : loco::all_nodes(g))
+  {
+    if (auto circle_dequant = dynamic_cast<luci::CircleDequantize *>(node))
+    {
+      auto input_node = circle_dequant->input();
+      if (auto const_input = dynamic_cast<luci::CircleConst *>(input_node))
+      {
+        // Input of Dequantize is constant
+        loco::replace(circle_dequant).with(dequantized_const_node(const_input));
+        changed = true;
+      }
+      else if (loco::dtype_get(input_node) == loco::DataType::FLOAT32)
+      {
+        // Input is already dequantized
+        loco::replace(circle_dequant).with(input_node);
+        changed = true;
+      }
+    }
+    else if (auto const_node = dynamic_cast<luci::CircleConst *>(node))
+    {
+      if (const_node->quantparam() != nullptr && is_foldable_const(const_node))
+      {
+        for (auto s : loco::succs(const_node))
+        {
+          auto s_user = loco::succs(s);
+          if (!is_hybrid_kernel_supported(s) && s_user.size() == 1 &&
+              dynamic_cast<luci::CircleDequantize *>(*s_user.begin()) != nullptr)
+          {
+            // Input can be pre-dequantized
+            changed |= replace_const_node(s, dequantized_const_node(const_node));
```

Comment:
This leaves dequantize Op after replacing the const_node. It would be better to call `loco::replace(s_user).with(s)`.

Reply:
I was confused with other pattern, thanks :)